src/main/PresentationExchange.ts

```diff
@@ -138,47 +137,70 @@ export class PresentationExchange {
    * returns a PresentationDefinition if a valid instance found
    * @param obj: object that can have a presentation_definition inside
    */
-  public static findValidPresentationDefinitions(obj: JWTPayload): PresentationDefinitionWithLocation[] {
+  public static async findValidPresentationDefinitions(obj: JWTPayload): Promise<PresentationDefinitionWithLocation[]> {
     let allDefinitions: PresentationDefinitionWithLocation[];
-
-    function extractPDFromVPToken() {
+    async function extractPDFromVPToken() {
       const vpTokens = extractDataFromPath(obj, '$..vp_token.presentation_definition');
+      const vpTokenRef = extractDataFromPath(obj, '$..vp_token.presentation_definition_uri');
+      if (vpTokens && vpTokens.length && vpTokenRef && vpTokenRef.length) {
+        throw new Error(SIOPErrors.REQUEST_CLAIMS_CANT_SEND_PRESENTATION_DEFINITION_BY_RE_AND_VAL);
+      }
+      if (vpTokens && vpTokens.length == 1) {
+        PresentationExchange.assertValidPresentationDefinition(vpTokens[0].value);
+        allDefinitions = [{ definition: vpTokens[0].value, location: PresentationLocation.VP_TOKEN }];
+      }
+      if (vpTokens && vpTokens.length > 1) {
+        throw new Error(SIOPErrors.REQUEST_CLAIMS_PRESENTATION_DEFINITION_NOT_VALID);
+      }
+      if (vpTokenRef && vpTokenRef.length) {
+        const result = await getWithUrl(vpTokenRef[0].value);
+        const pd: PresentationDefinitionV1 | PresentationDefinitionV2 = await result.json();
+        PresentationExchange.assertValidPresentationDefinition(pd);
+        allDefinitions = [{ definition: pd, location: PresentationLocation.VP_TOKEN }];
+      }
+    }
 
-      if (vpTokens) {
-        if (vpTokens.length == 1) {
-          PresentationExchange.assertValidPresentationDefinition(vpTokens[0].value);
-          allDefinitions = [{ definition: vpTokens[0].value, location: PresentationLocation.VP_TOKEN }];
-        } else if (vpTokens.length > 1) throw new Error(SIOPErrors.REQUEST_CLAIMS_PRESENTATION_DEFINITION_NOT_VALID);
+    function handleAddingSinglePDInIdToken(definition, allDefinitions: PresentationDefinitionWithLocation[]): PresentationDefinitionWithLocation[] {
```

Comment:
addSingleIdTokenPDToPDs to make it a bit more clear that the definition is being added. Actually to make it more clear the input array is modified this probably should return a void

Reply:
changed